### PR TITLE
feat: Add smart document chunking that splits by topic

### PR DESCRIPTION
**Pull request by tobocop2**
## What this does

Adds a `Semantic` chunker type that splits documents at natural topic boundaries instead of fixed character windows. Each chunk stays focused on one subject.

## Usage

```python
# Works out of the box - no configuration needed
config = ChunkingConfig(chunker_type="semantic")

# With embedding-based topic detection
config = ChunkingConfig(
    chunker_type="semantic",
    embedding=EmbeddingConfig(model=EmbeddingModelType.preset("fast")),
)
```

Available in all 12 language bindings with `chunker_type: "semantic"`.

## How it works

- Splits text into paragraph-level segments
- Detects structural boundaries (ALL CAPS headers, numbered sections, title lines)
- Optionally compares adjacent segments by cosine similarity when an embedding model is configured
- Each topic group becomes one chunk. Oversized groups are split at a safety ceiling derived from the embedding model's context window (or 4000 chars when no model is configured)

The chunk budget is derived from the document's structure, not from a fixed `max_characters` value. Callers don't need to think about chunk sizes.

## Semantic chunker changes

- New modules: `boundary_detection`, `semantic/mod`, `semantic/merge`, `semantic/topic`
- Config: `ChunkerType::Semantic` variant, optional `topic_threshold` field (default 0.75)
- All 12 bindings, CLI, API server, MCP server updated
- E2e contract fixtures and test documents added
- Reference docs, API guide, CLI README updated
- `preset_chunk_size()` helper added to embeddings module

## Shared utility extraction

- Extract `is_markdown_header` into `utils/markdown_utils` as a shared utility with no feature gate
- Both `chunking` and `quality` now share the same implementation
- Tests consolidated in `utils/markdown_utils`

Implements https://github.com/kreuzberg-dev/kreuzberg/issues/728

## CI notes

The tree-sitter e2e tests (`code_python_basic`, `code_tree_sitter_config`, `config_tree_sitter`, `config_tree_sitter_process`) and R e2e tests fail in CI but are pre-existing on main. This PR does not modify any e2e test files, tree-sitter code, or R package code. The `Validate` and `Validate: Rust` steps both pass.

### CODE REVIEW

File: crates/kreuzberg/src/utils/markdown_utils.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+//! Lightweight markdown detection utilities.
```
Comment (tobocop2) [inline crates/kreuzberg/src/utils/markdown_utils.rs:1]: i know this is a little out of scope, but I moved this functionality here so that it could be reusable between the token reduction and chunking tools

File: crates/kreuzberg/src/chunking/semantic/merge.rs [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+        }
+    }
+    groups.push(group_start..segments.len());
+
+    let mut result: Vec<MergedChunk> = Vec::new();
+    let mut prev_group_tail: Option<String> = None;
+
+    for group in &groups {
+        let group_segments = &segments[group.clone()];
+        let group_text: String = group_segments.iter().map(|s| s.text).collect::<Vec<_>>().join("\n\n");
```
Comment (Goldziher) [inline crates/kreuzberg/src/chunking/semantic/merge.rs]: **CRITICAL: byte_start/byte_end invariant broken for merged segments**

This join creates a *new* string that does not exist at any position in the original `text`:
```rust
let group_text = group_segments.iter().map(|s| s.text).collect::<Vec<_>>().join("\n\n");
```

But `byte_start`/`byte_end` (lines 106-108) are taken from the *original* segment offsets. Downstream consumers that do `&original_text[byte_start..byte_end]` will get content that does not match `chunk.content`.

This breaks the contract tested by `test_issue_439_chunk_byte_offsets_match_text_position` in `core.rs`.

**Fix options:**
1. Use the original text between segments (`&text[first_seg.byte_start..last_seg_byte_end]`) instead of joining with `"\n\n"`, preserving the true subslice relationship
2. Recompute byte offsets relative to the new concatenated string (but this breaks page attribution)

File: crates/kreuzberg/src/chunking/semantic/topic.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        }
+    }
+
+    if n < 2 {
+        return Ok(boundaries);
+    }
+
+    let embeddings = crate::embeddings::embed_texts(segment_texts, embedding_config)?;
+
+    for i in 1..n {
```
Comment (Goldziher) [inline crates/kreuzberg/src/chunking/semantic/topic.rs:80]: **HIGH: No bounds check on embeddings return**

If `embed_texts` returns fewer embeddings than segments (buggy backend, partial failure, future API change), `embeddings[i]` panics with index out of bounds.

Add a length assertion after the `embed_texts` call:
```rust
if embeddings.len() != n {
    return Err(format!("expected {} embeddings, got {}", n, embeddings.len()).into());
}
```

File: crates/kreuzberg/src/chunking/boundary_detection.rs [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+
+    true
+}
+
+/// Check if a line looks like a standalone title.
+///
+/// Criteria: < 60 chars, starts with uppercase letter, title-case words
+/// (most words start with uppercase), no more than 8 words, no trailing
+/// sentence punctuation (. ! ? ; :), not indented.
+fn is_title_line(line: &str) -> bool {
```
Comment (Goldziher) [inline crates/kreuzberg/src/chunking/boundary_detection.rs:132]: **HIGH: `is_title_line` false positives on ordinary prose**

Single capitalised words pass unconditionally — "Note", "I", "To", "Figure 3", "Table 2" all trigger false boundary splits. Any sentence <=60 chars without terminal `.!?;:` that starts uppercase and has >=60% title-cased words is also accepted.

Consider:
- Require minimum word count of 2
- Require the line to appear between blank lines (context the function currently discards)
- Or remove `is_title_line` entirely and rely on the more reliable ALL_CAPS and numbered-section heuristics

File: crates/kreuzberg/src/chunking/semantic/mod.rs [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    };
+
+    if splitter_segments.is_empty() {
+        return Ok(ChunkingResult {
+            chunks: vec![],
+            chunk_count: 0,
+        });
+    }
+
+    let source_start = text.as_ptr() as usize;
```
Comment (Goldziher) [inline crates/kreuzberg/src/chunking/semantic/mod.rs:66]: **HIGH: Pointer arithmetic relies on undocumented text_splitter contract**

```rust
let source_start = text.as_ptr() as usize;
byte_start: s.as_ptr() as usize - source_start,
```

This assumes `text_splitter::chunks()` returns subslices of the original text. This is true today but is not a documented API guarantee. A version bump could silently break this.

At minimum add a bounds assertion:
```rust
debug_assert!(s.as_ptr() >= text.as_ptr()
    && s.as_ptr() as usize + s.len() <= text.as_ptr() as usize + text.len(),
    "text_splitter segment is not a subslice of the input");
```

File: crates/kreuzberg/src/chunking/semantic/mod.rs [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+            text: s,
+            byte_start: s.as_ptr() as usize - source_start,
+        })
+        .collect();
+
+    let detected = detect_plain_text_boundaries(text);
+    let mut forced: Vec<bool> = vec![false; segments.len()];
+    // First segment always starts a group.
+    forced[0] = true;
+    for boundary in &detected {
```
Comment (Goldziher) [inline crates/kreuzberg/src/chunking/semantic/mod.rs:86]: **MEDIUM: O(n*m) boundary-to-segment matching**

Nested loop over `detected` boundaries x `segments`. Both are sorted by byte offset, so this should be an O(n+m) two-pointer merge:

```rust
let mut seg_idx = 0;
for boundary in &detected {
    while seg_idx < segments.len() && segments[seg_idx].byte_start < boundary.byte_offset {
        seg_idx += 1;
    }
    if seg_idx < segments.len() { forced[seg_idx] = true; }
}
```

File: crates/kreuzberg/src/chunking/boundary_detection.rs [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+    if line.starts_with(' ') || line.starts_with('\t') {
+        return false;
+    }
+
+    let trimmed = line.trim();
+    if trimmed.is_empty() {
+        return false;
+    }
+
+    // Find the numbering prefix: digits and dots
```
Comment (Goldziher) [inline crates/kreuzberg/src/chunking/boundary_detection.rs:89]: **MEDIUM: Bare numeric prefix accepted as section header**

A line containing only `"1."` or `"1.2.3."` with no following text satisfies all conditions. This will fire on footnote references, decimal numbers at line start in tables, etc.

Require at least one non-whitespace character after the numeric prefix.

File: crates/kreuzberg/src/core/config/processing.rs [hunk 8]
```diff
@@ -167,6 +173,16 @@ pub struct ChunkingConfig {
...
+
+    /// Optional cosine similarity threshold for semantic topic boundary detection.
+    ///
+    /// Only used when `chunker_type` is `Semantic` and an `EmbeddingConfig` is
+    /// provided. You almost never need to set this. When omitted, defaults to
+    /// `0.75` which works well for most documents. Lower values detect more
+    /// topic boundaries (more, smaller chunks); higher values detect fewer.
+    /// Range: `0.0..=1.0`.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub topic_threshold: Option<f32>,
```
Comment (Goldziher) [inline crates/kreuzberg/src/core/config/processing.rs:185]: **MEDIUM: `topic_threshold` not validated**

No validation that the value is in `[0.0, 1.0]`. A value of 2.0 means cosine similarity (max 1.0) never exceeds the threshold, silently producing a single chunk for the entire document. A value of -1.0 marks every segment as a boundary.

Add validation in the config constructor or at the `compute_boundaries` call site.

File: crates/kreuzberg-wasm/typescript/types.ts [hunk 9]
```diff
@@ -605,7 +605,9 @@ export interface ChunkingConfig {
 	maxOverlap?: number;
 	/** Named preset for chunking strategy (e.g., "balanced", "fast", "semantic") */
 	preset?: string;
-	/** Chunker type: "text" (default), "markdown", or "yaml" */
+	/** Chunker type: "text" (default), "markdown", "yaml", or "semantic".
+	 * Set to "semantic" for topic-aware chunking that works out of the box
+	 * with sensible defaults. No other parameters needed. */
 	chunkerType?: string;
```
Comment (Goldziher) [inline crates/kreuzberg-wasm/typescript/types.ts]: **MEDIUM: `chunkerType` typed as bare `string`**

The TypeScript npm package uses a literal union for type safety. This should match:
```typescript
chunkerType?: "text" | "markdown" | "yaml" | "semantic";
```

**Review by Goldziher — CHANGES_REQUESTED** [review]
## Review Summary

17 issues found across correctness, binding parity, docs, and testing.

### Critical (1)
- **Byte offset invariant broken** — merged segments create new allocations via `"\n\n"` join but `byte_start`/`byte_end` still reference the original text. Breaks page attribution, heading context, and the existing `test_issue_439_chunk_byte_offsets_match_text_position` contract. See inline comment on `merge.rs:103`.

### High (6)
- **`topic_threshold` unreachable from API, MCP, CLI** — `ChunkingConfigRequest` (`api/types.rs:273`), `ChunkTextParams` (`mcp/params.rs:147`), and the CLI chunk subcommand all lack a `topic_threshold` field/flag. Users always get the hardcoded 0.75. `ChunkingConfigResponse` also doesn't echo it back.
- **Elixir binding not updated** — `packages/elixir/lib/kreuzberg/config.ex` `validate_chunking_config/1` has no awareness of `chunker_type` or `topic_threshold`. PR claims "all 12 bindings".
- **R binding not updated** — `packages/r/R/config.R` `chunking_config()` only accepts `max_characters`, `overlap`, and `...` passthrough. No documented parameters for `chunker_type` or `topic_threshold`.
- **No bounds check on embeddings return** — see inline on `topic.rs:72`
- **`is_title_line` false positives** — see inline on `boundary_detection.rs:125`
- **Pointer arithmetic for byte_start** — see inline on `mod.rs:66`

### Medium (6)
- **`topic_threshold` not validated** to `[0.0, 1.0]` — see inline on `processing.rs:185`
- **O(n*m) boundary-to-segment matching** — see inline on `mod.rs:79`
- **Bare numeric prefix** `"1."` with no title text accepted as section header — see inline on `boundary_detection.rs:89`
- **Python `__repr__`** missing `chunker_type` and `topic_threshold` (`kreuzberg-py/src/config/types.rs:~1110`)
- **TypeScript npm JSDoc** on `chunkerType` still says `"text" or "markdown"` — missing `"yaml"` and `"semantic"` (`packages/typescript/core/src/types/config.ts:~118`)
- **WASM `chunkerType`** typed as bare `string` instead of literal union — see inline on `types.ts:611`

### Docs & Testing (4)
- **E2E tests not generated** — the 3 semantic fixtures exist (`fixtures/contract/config_chunking_semantic*.json`) but `task generate:e2e` was not run. No language has generated tests for these fixtures yet. Please regenerate and commit.
- **CHANGELOG.md** — no entry in `[Unreleased]` section. Add an `### Added` subsection for the semantic chunker.
- **docs/guides/advanced.md** line 5 — still says "Two strategies: **Text** and **Markdown**". Needs **Semantic** as a third strategy with a section explaining usage.
- **Code snippets** — `docs/snippets/` chunking examples only show "basic" and "markdown". Need semantic examples for at least Python, TypeScript, and Rust.

See inline comments for detailed explanations and fix suggestions.

File: crates/kreuzberg/src/utils/markdown_utils.rs (hunk 1 above)
Comment (tobocop2) [reply crates/kreuzberg/src/utils/markdown_utils.rs:1]: Done.

File: crates/kreuzberg/src/chunking/semantic/merge.rs (hunk 2 above)
Comment (tobocop2) [reply crates/kreuzberg/src/chunking/semantic/merge.rs]: Fixed. merge_segments now takes source_text and slices the original text instead of joining with "\\n\\n". Byte offsets are now valid subslice references.

File: crates/kreuzberg/src/chunking/semantic/topic.rs (hunk 3 above)
Comment (tobocop2) [reply crates/kreuzberg/src/chunking/semantic/topic.rs:80]: Done. Added length validation after embed_texts call, returns KreuzbergError::validation on mismatch.

File: crates/kreuzberg/src/chunking/boundary_detection.rs (hunk 4 above)
Comment (tobocop2) [reply crates/kreuzberg/src/chunking/boundary_detection.rs:132]: Fixed. Now requires minimum 2 words. Single capitalized words like "Note", "I", "To" are rejected.

File: crates/kreuzberg/src/chunking/semantic/mod.rs (hunk 5 above)
Comment (tobocop2) [reply crates/kreuzberg/src/chunking/semantic/mod.rs:66]: Done. Added debug_assert verifying each segment is a valid subslice of the input text.

File: crates/kreuzberg/src/chunking/semantic/mod.rs (hunk 6 above)
Comment (tobocop2) [reply crates/kreuzberg/src/chunking/semantic/mod.rs:86]: Fixed. Replaced with O(n+m) two-pointer merge.

File: crates/kreuzberg/src/chunking/boundary_detection.rs (hunk 7 above)
Comment (tobocop2) [reply crates/kreuzberg/src/chunking/boundary_detection.rs:89]: Fixed. Bare prefixes like "1." with no following text are now rejected. Added test cases.

File: crates/kreuzberg/src/core/config/processing.rs (hunk 8 above)
Comment (tobocop2) [reply crates/kreuzberg/src/core/config/processing.rs:185]: Done. Builder asserts [0.0, 1.0] range. Runtime clamps deserialized values. Added test cases for boundary values and out-of-range panics.

File: crates/kreuzberg-wasm/typescript/types.ts (hunk 9 above)
Comment (tobocop2) [reply crates/kreuzberg-wasm/typescript/types.ts]: Done. Both TypeScript core and WASM now use literal union type.

**Comment by tobocop2** [comment]
All 17 items addressed:

**Critical:** Byte offset invariant fixed - `merge_segments` now slices the original source text instead of joining with `"\n\n"`.

**High:**
- `topic_threshold` added to API (`ChunkingConfigRequest` + `ChunkingConfigResponse`), MCP (`ChunkTextParams`), and CLI (`--topic-threshold` flag). Sorry for the gap - should have traced all entry points.
- Elixir: added `validate_chunker_type/1` and `validate_topic_threshold/1` to `validate_chunking_config/1`.
- R: added explicit `chunker_type` and `topic_threshold` parameters with validation.
- Embeddings bounds check added after `embed_texts` call.
- `is_title_line` now requires minimum 2 words.
- Pointer arithmetic assertion added.

**Medium:**
- `topic_threshold` validated to `[0.0, 1.0]` in builder (assert) and at use site (clamp).
- O(n*m) replaced with O(n+m) two-pointer merge.
- Bare numeric prefix without following text now rejected.
- Python `__repr__` includes `chunker_type` and `topic_threshold`.
- TypeScript and WASM `chunkerType` use literal union type.

**Docs & Testing:**
- CHANGELOG entry added.
- `docs/guides/advanced.md` updated with Semantic strategy section.
- Code snippets added for Python, Rust, TypeScript.
- E2e tests regenerated across all 10 languages from fixtures.

**Comment by Goldziher** [comment]
@tobocop2 - please rebase. regenerate e2e tests when done, and run `prek run --all-files`

**Review by Goldziher — APPROVED** [review]
merge when ready @tobocop2
